pkg/cli/version/version.go

```diff
@@ -1,25 +1,186 @@
+		# Print the OpenShift client, Kubernetes server, and OpenShift server version information for the current context.
+		%[1]s version
+	    # Print the OpenShift client, Kubernetes server, and OpenShift server version numbers for the current context.
+	    %[1]s version --short
+		# Print the OpenShift client version information for the current context.
+		%[1]s version --client`)
+)
+
+type VersionOptions struct {
+	oClient         configv1client.ClusterOperatorsGetter
+	discoveryClient discovery.CachedDiscoveryInterface
+
+	genericclioptions.IOStreams
+}
+
+func NewVersionOptions(ioStreams genericclioptions.IOStreams) *VersionOptions {
+	return &VersionOptions{
+		IOStreams: ioStreams,
+	}
+}
+
+// NewCmdVersion is copied from upstream NewCmdVersion with added "server" flag
+func NewCmdVersion(fullName string, f cmdutil.Factory, ioStreams genericclioptions.IOStreams) *cobra.Command {
+	o := NewVersionOptions(ioStreams)
+	ko := kversion.NewVersionOptions(ioStreams)
 	cmd := &cobra.Command{
-		Use:   "version",
-		Short: "Display version",
-		Long:  "Display version",
+		Use:     "version",
+		Short:   i18n.T("Print the client and server version information"),
+		Long:    "Print the client and server version information for the current context",
+		Example: fmt.Sprintf(versionExample, fullName),
 		Run: func(cmd *cobra.Command, args []string) {
-			fmt.Fprintf(out, "%s %v\n", fullName, versionInfo)
+			cmdutil.CheckErr(ko.Validate())
+			cmdutil.CheckErr(o.Complete(f, cmd, ko))
+			cmdutil.CheckErr(o.Run(ko))
 		},
 	}
-
+	cmd.Flags().BoolVar(&ko.ClientOnly, "client", ko.ClientOnly, "Client version only (no server required).")
+	cmd.Flags().BoolVar(&ko.Short, "short", ko.Short, "Print just the version number.")
+	cmd.Flags().StringVarP(&ko.Output, "output", "o", ko.Output, "One of 'yaml' or 'json'.")
 	return cmd
 }
+
+// Complete is copied from upstream version command with added clusteroperator client
+// to report OpenShift server version
+func (o *VersionOptions) Complete(f cmdutil.Factory, cmd *cobra.Command, ko *kversion.VersionOptions) error {
+	var err error
+	if ko.ClientOnly {
+		return nil
+	}
+	o.discoveryClient, err = f.ToDiscoveryClient()
+	// if we had an empty rest.Config, continue and just print out client information.
+	// if we had an error other than being unable to build a rest.Config, fail.
+	if err != nil && !clientcmd.IsEmptyConfig(err) {
+		return err
+	}
+
+	clientConfig, err := f.ToRESTConfig()
+	if err != nil && !clientcmd.IsEmptyConfig(err) {
+		return err
+	}
+	if clientConfig != nil {
+		o.oClient, err = configv1client.NewForConfig(clientConfig)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Run is copied from upstream version command, with added OpenShift server version logic
+func (o *VersionOptions) Run(ko *kversion.VersionOptions) error {
+	var (
+		serverErr       error
+		serverVersion   *apimachineryversion.Info
+		clusterOperator *configv1.ClusterOperator
+		versionInfo     Version
```

Comment:
Move these directly where they are used. It's uncommon in Go code to have these at the topc of the func.

Reply:
moved them